[#110149] - condition if imagePullPolicy is set to Always

dariodsa opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:
If ``imagePullPolicy`` is set to ``Always``, ``image_locality`` plugin should score it with zero points, because image will be either way pulled. To sum up, previously ``image_locality`` scored nodes, not taking in the consideration the ``ImagePullPolicy``. This PR fixed this issue. 

https://github.com/kubernetes/kubernetes/blob/9af00439cc430e7e643bb2357de619da7d5975cc/pkg/scheduler/framework/plugins/imagelocality/image_locality.go#L100-L103


#### Which issue(s) this PR fixes:

Fixes #110149 

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

dariodsa committed:
[#110149] - condition if imagePullPolicy is set to Always

yxxhero commented:
@dariodsa Add some unittest for this?

dariodsa commented:
> @dariodsa Add some unittest for this?

Sorry, ``image_locality_test.go`` is updated.

damemi commented on pkg/scheduler/framework/plugins/imagelocality/image_locality.go:
```diff
@@ -97,6 +97,10 @@ func calculatePriority(sumScores int64, numContainers int) int64 {
 	var sum int64
 	for _, container := range containers {
+		// if pullPolicy is set to Always score it with 0 points
+		if container.ImagePullPolicy == "Always" {
```
could this just fail earlier in the plugin, before needing to list all the nodes? if we know the container has an `Always`, then do we need to loop through the rest of the nodes? I suppose it will still have a scheduling effect for the other containers, so maybe we do need to do it this way. Just curious if it was considered

dariodsa replied:
True, I didnt consider it. We can have list of a containers which doesnt
have imagePullPolicy set to Always. Others will ve calculated normally and
one's with Always will either way always have 0 score so that doesnt matter.
I will update code later this day.

damemi reviewed (commented):
/ok-to-test

dariodsa commented:
True, I didnt consider it. We can have list of a containers which doesnt
have imagePullPolicy set to Always. Others will ve calculated normally and
one's with Always will either way always have 0 score so that doesnt matter.
I will update code later this day.

On Mon, 6 Jun 2022, 16:33 Mike Dame, ***@***.***> wrote:

> ***@***.**** commented on this pull request.
>
> /ok-to-test
> ------------------------------
>
> In pkg/scheduler/framework/plugins/imagelocality/image_locality.go
> <https://github.com/kubernetes/kubernetes/pull/110382#discussion_r890214505>
> :
>
> > @@ -97,6 +97,10 @@ func calculatePriority(sumScores int64, numContainers int) int64 {
>  func sumImageScores(nodeInfo *framework.NodeInfo, containers []v1.Container, totalNumNodes int) int64 {
>  	var sum int64
>  	for _, container := range containers {
> +		// if pullPolicy is set to Always score it with 0 points
> +		if container.ImagePullPolicy == "Always" {
>
> could this just fail earlier in the plugin, before needing to list all the
> nodes? if we know the container has an Always, then do we need to loop
> through the rest of the nodes? I suppose it will still have a scheduling
> effect for the other containers, so maybe we do need to do it this way.
> Just curious if it was considered
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/110382#pullrequestreview-996709518>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA7DGPU467QYHWT6KEV2GETVNYD43ANCNFSM5X3ICEBA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

dariodsa commented:
I updated code so now I first perform filtering on ImagePullPolicy and pass that filtered list to the function. Containers with Always pullPolicy will either way score 0 points so we doesn't need to pass it to the function. 
https://github.com/kubernetes/kubernetes/blob/141de39419478d0a8850c2ab1f61e501270ad571/pkg/scheduler/framework/plugins/imagelocality/image_locality.go#L66-L75

damemi commented on pkg/scheduler/framework/plugins/imagelocality/image_locality.go:
```diff
@@ -63,7 +63,16 @@ func (pl *ImageLocality) Score(ctx context.Context, state *framework.CycleState,
-	score := calculatePriority(sumImageScores(nodeInfo, pod.Spec.Containers, totalNumNodes), len(pod.Spec.Containers))
+	// Filtering containers with ImagePullPolicy different from Always
+	// Ones with Always will score 0 point either way
+	var notAlwaysContainers []v1.Container;
+	for _, container := range pod.Spec.Containers {
+		if container.ImagePullPolicy != "Always" {
+			notAlwaysContainers = append(notAlwaysContainers, container)
+		}
+	}
```
in the worst case, where none of the containers have `Always`, we'll end up looping through the whole list twice (because [`sumImageScores`](https://github.com/kubernetes/kubernetes/blob/46f3e4d/pkg/scheduler/framework/plugins/imagelocality/image_locality.go#L97-L105) also loops through all the containers)

So what about updating the [check in `sumImageScores`](https://github.com/kubernetes/kubernetes/blob/46f3e4d/pkg/scheduler/framework/plugins/imagelocality/image_locality.go#L100) to just:
```go
if state, ok := nodeInfo.ImageStates[normalizedImageName(container.Image)]; ok && container.ImagePullPolicy != "Always" {
...
}
```

damemi replied:
Oh wow, I'm sorry, that's basically exactly what you had before I asked you to change it 🤦

damemi reviewed (commented):
one more nit, but the idea looks good to me
just want to ping @alculquicondor to make sure there's no scenarios I'm missing where this could have an adverse effect

alculquicondor commented:
> because image will be either way pulled

This is not 100% true. kubelet will indeed issue an image pull request down to the CRI, but the CRI might just use its internal cache, if the image didn't change in the image repository.

alculquicondor commented on pkg/scheduler/framework/plugins/imagelocality/image_locality.go:
```diff
@@ -63,7 +63,16 @@ func (pl *ImageLocality) Score(ctx context.Context, state *framework.CycleState,
+	// Ones with Always will score 0 point either way
+	var notAlwaysContainers []v1.Container;
+	for _, container := range pod.Spec.Containers {
+		if container.ImagePullPolicy != "Always" {
```
```suggestion
		if container.ImagePullPolicy != v1.PullAlways {
```

alculquicondor reviewed (commented):
/approve

alculquicondor commented:
Please fix the format and spelling https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/110382/pull-kubernetes-verify/1533904763643498496

Huang-Wei reviewed (commented):
/hold

Generally, the idea looks good to me. However, I'd like to place the logic of checking "whether a pod should go through ImageLocality" to its (new) PreScore function; otherwise the check occurs on every filtered node duplicately.

BTW: you can refer to other plugin's PreScore implementation. It usually plumbs a key:value pair to the CycleState and read in Score phase.

dariodsa commented:
Taking in consideration all feedbacks, I come with this implementation. There were a lot of changes, so let me go through them. 
After adding all necessary stuff for ``PreScore`` like ``preScoreState``, ``preScoreName`` and etc I implemented ``PreScore`` function which updates the ``preScoreState`` with all containers which doesn't have ``imagePullPolicy`` ``Always``.
https://github.com/kubernetes/kubernetes/blob/3b964cbda0e66c47a22a70ee4771b713ba0d2835/pkg/scheduler/framework/plugins/imagelocality/image_locality.go#L105-L122
So in ``Score`` function I get preScoreState and check if that container list is empty, if it is then I return 0 immediately and I don't go over the nodes. Only containers with ImagePullPolicy different from Always are passed to the ``calculatePriority`` function so in that sense number of containers in that part is minimized. 
https://github.com/kubernetes/kubernetes/blob/3b964cbda0e66c47a22a70ee4771b713ba0d2835/pkg/scheduler/framework/plugins/imagelocality/image_locality.go#L85-L89
I also updated the test function to call preScore as well.

dariodsa commented:
I still have issues regarding adding ``ImageLocality`` in preScore in ``v1beta3``. I can easily add it in ``v1beta2`` but I don't know where to add it to ``v1beta3``. Can someone advice me on this?

https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/apis/config/v1beta3/default_plugins.go#L28
https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/apis/config/v1beta2/default_plugins.go#L30

kerthcet commented:
I think test failed for you forget to update the pluginsets in tests. Notice that we instantiated another sets of plugins only for tests, see here  https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/apis/config/testing/defaults/defaults.go

Huang-Wei commented on pkg/scheduler/framework/plugins/imagelocality/image_locality.go:
```diff
@@ -73,6 +96,31 @@ func (pl *ImageLocality) ScoreExtensions() framework.ScoreExtensions {
+}
+
+func (pl *ImageLocality) PreScore(ctx context.Context, cycleState *framework.CycleState, pod *v1.Pod, nodes []*v1.Node) *framework.Status {
+	var containersPullAlways []v1.Container;
```
Go code doesn't end with `;`

Huang-Wei commented on pkg/scheduler/framework/plugins/imagelocality/image_locality.go:
```diff
@@ -42,8 +42,17 @@ type ImageLocality struct {
+
+// preScoreState computed at PreScore and used at Score.
+type preScoreState struct {
+	containersPullAlways []v1.Container
```
do we need a full list of containers? or just a single bool variable to indicate whether all containers are using PullAlways.

dariodsa replied:
I choose list of containers because I didn't want to loop over containers with pullPolicy set to Always twice. So ``preScoreState`` list is filled with containers on which ImageLocality score has to be calculated (other than Always).  I mean we can shrink it to a list of a strings ( just image property).

alculquicondor commented on pkg/scheduler/framework/plugins/imagelocality/image_locality.go:
```diff
@@ -63,7 +72,21 @@ func (pl *ImageLocality) Score(ctx context.Context, state *framework.CycleState,
 	totalNumNodes := len(nodeInfos)
 
-	score := calculatePriority(sumImageScores(nodeInfo, pod.Spec.Containers, totalNumNodes), len(pod.Spec.Containers))
+	c, err := state.Read(preScoreStateKey)
```
In general, you should assume that PreScore might be disabled, in which case you should run the logic to get all the images in Score.

dariodsa replied:
https://github.com/kubernetes/kubernetes/blob/06cf52262796f386260beab3bfb4d0a29ab7b665/pkg/scheduler/framework/plugins/imagelocality/image_locality.go#L77-L78
I didn't see that other plugin do that, but they should. I only need to do hack with creating empty node list, but it isn't need in preScore function anyway. It is here only because of the interface.

alculquicondor replied:
Can you follow a similar approach? That is, have a unexported method that just knows how to get the data without writing it to the state.

Huang-Wei commented on pkg/scheduler/framework/plugins/imagelocality/image_locality.go:
```diff
@@ -63,7 +72,21 @@ func (pl *ImageLocality) Score(ctx context.Context, state *framework.CycleState,
+		return 0, nil;
+	}
+
+	score := calculatePriority(sumImageScores(nodeInfo, s.containersPullAlways, totalNumNodes), len(pod.Spec.Containers))
```
oh, just realized that we used to only calculate regular containers. However, I think initContainers should also be counted. @alculquicondor wdyt?

alculquicondor replied:
I guess we could take them into consideration, but that would be a separate issue/PR

kerthcet replied:
Considering init containers start in sequence, I think we should only pick the max value.

kerthcet commented on pkg/scheduler/framework/plugins/imagelocality/image_locality.go:
```diff
@@ -42,8 +42,17 @@ type ImageLocality struct {
 
 var _ framework.ScorePlugin = &ImageLocality{}
 
```
remove the blank line.

kerthcet commented on pkg/scheduler/framework/plugins/imagelocality/image_locality.go:
```diff
@@ -63,7 +72,22 @@ func (pl *ImageLocality) Score(ctx context.Context, state *framework.CycleState,
+
+	s, ok := c.(*preScoreState)
+	if !ok {
+		return 0, framework.AsStatus(fmt.Errorf("cannot convert saved state to selectorspread.preScoreState"))
```
s/selectorspread/imageLocality/

kerthcet commented on pkg/scheduler/framework/plugins/imagelocality/image_locality.go:
```diff
@@ -73,6 +97,31 @@ func (pl *ImageLocality) ScoreExtensions() framework.ScoreExtensions {
+	// Ones with Always will score 0 point either way
+	for _, container := range pod.Spec.Containers {
+		if container.ImagePullPolicy != v1.PullAlways {
+			containersPullAlways = append(containersPullAlways, container)
```
Up to now, `container.Image` is enough, let's use a `sets.String` instead which helps to reduce the memory allocation?

kerthcet commented on pkg/scheduler/framework/plugins/imagelocality/image_locality.go:
```diff
@@ -73,6 +97,31 @@ func (pl *ImageLocality) ScoreExtensions() framework.ScoreExtensions {
+	return s
+}
+
+func (pl *ImageLocality) PreScore(ctx context.Context, cycleState *framework.CycleState, pod *v1.Pod, nodes []*v1.Node) *framework.Status {
```
What's the nodes use for?

kerthcet commented on pkg/scheduler/framework/plugins/imagelocality/image_locality_test.go:
```diff
@@ -328,20 +368,79 @@ func TestImageLocalityPriority(t *testing.T) {
+			pod:          &v1.Pod{Spec: test3040},
+			nodes:        []*v1.Node{makeImageNode("node1", node203040), makeImageNode("node2", node400030)},
+			expectedList: []framework.NodeScore{{Name: "node1", Score: 1}, {Name: "node2", Score: 0}},
+			name:         "pod with multiple small images",
```
What's the difference between this one and the test case above? Besides, better to use different test case names for each case.

kerthcet replied:
Same question here?

alculquicondor commented:
@dariodsa are you still working on this?

xiaomudk commented:
Let me try

dariodsa commented:
> Let me try

@xiamudk, you can continue working on this issue.

kerthcet commented:
@xiaomudk Are you still working on this?

kerthcet commented:
/remove-lifecycle rotten

xiaomudk commented:
@kerthcet Yes, please give me some time, I will finish it as soon as possible

alculquicondor commented:
Please rebase

dims commented:
If you still need this PR then please rebase, if not, please close the PR

xiaomudk commented:
/assign

xiaomudk commented:
@dariodsa   Is it possible for you to add me as a collaborator on your fork of the repository?

dariodsa commented:
@xiaomudk I added you on my repo as collaborator.

dariodsa commented:
What is status?

xiaomudk commented:
@dariodsa I am working on it.

xiaomudk commented:
/cc @damemi @kerthcet @alculquicondor

xiaomudk commented:
/retest-required

alculquicondor commented:
@kerthcet can you take another look before approval?

kerthcet commented:
> @kerthcet can you take another look before approval?

/assign

kerthcet commented on pkg/scheduler/framework/plugins/imagelocality/image_locality.go:
```diff
@@ -41,17 +41,67 @@ type ImageLocality struct {
 func (pl *ImageLocality) Score(ctx context.Context, state *framework.CycleState, pod *v1.Pod, nodeName string) (int64, *framework.Status) {
+	s, err := getPreScoreState(state)
+	if err != nil {
+		pl.PreScore(ctx, state, pod, []*v1.Node{})
```
Or maybe oneday preScore & Score should be enabled/disabled together.(same as preFilter & Filter)

kerthcet commented on pkg/scheduler/framework/plugins/imagelocality/image_locality.go:
```diff
@@ -41,17 +41,67 @@ type ImageLocality struct {
 }
 
+func (pl *ImageLocality) PreScore(ctx context.Context, cycleState *framework.CycleState, pod *v1.Pod, nodes []*v1.Node) *framework.Status {
+	pullAlways := true
+
+	// Filtering containers with ImagePullPolicy different from Always
+	// Ones with Always will score 0 point either way
+	for _, container := range pod.Spec.Containers {
+		if container.ImagePullPolicy != v1.PullAlways {
+			pullAlways = false
+			break
+		}
+	}
```
Also update the comment.

xiaomudk replied:
Only when all containers are PullAlways, the score will be 0. should it be？

kerthcet replied:
Rethink, it should be for containers whose imagePullPolicy == pullAlways, the score is 0, but for the other containers, we should still calculate the score, so a single boolean is not enough. we should at least have a sets of container names.

kerthcet commented on pkg/scheduler/framework/plugins/imagelocality/image_locality.go:
```diff
@@ -41,17 +41,67 @@ type ImageLocality struct {
+	s, err := getPreScoreState(state)
+	if err != nil {
+		pl.PreScore(ctx, state, pod, []*v1.Node{})
+		s, err = getPreScoreState(state)
```
Can we get the necessary information without writing to the state?  We can make finding the `imagePullPolicy=Always` container a method and recall here.

xiaomudk replied:
When PreScore is not enabled, only the first node needs to call PreScore to obtain cycleState, and other nodes can directly read cycleState. If it is changed to a function, each node needs to be executed once

kerthcet commented on pkg/scheduler/framework/plugins/imagelocality/image_locality.go:
```diff
@@ -97,7 +160,8 @@ func calculatePriority(sumScores int64, numContainers int) int64 {
 	for _, container := range containers {
-		if state, ok := nodeInfo.ImageStates[normalizedImageName(container.Image)]; ok {
+		state, ok := nodeInfo.ImageStates[normalizedImageName(container.Image)]
+		if ok && container.ImagePullPolicy != v1.PullAlways {
```
This is impossible, right? For we'll early return in Score() if we have containers with `imagePullPolicy=Always`.

xiaomudk replied:
see: https://github.com/kubernetes/kubernetes/pull/110382#discussion_r1096052515

kerthcet commented:
kind ping @xiaomudk for feedbacks.

xiaomudk commented on pkg/scheduler/framework/plugins/imagelocality/image_locality.go:
```diff
@@ -41,17 +41,67 @@ type ImageLocality struct {
+
+// preScoreState computed at PreScore and used at Score.
+type preScoreState struct {
+	containersPullAlways bool
```
see: https://github.com/kubernetes/kubernetes/pull/110382#discussion_r1096052515

xiaomudk commented on pkg/scheduler/framework/plugins/imagelocality/image_locality.go:
```diff
@@ -41,17 +41,67 @@ type ImageLocality struct {
 }
 
 var _ framework.ScorePlugin = &ImageLocality{}
+var _ framework.PreScorePlugin = &ImageLocality{}
```
done

xiaomudk commented on pkg/scheduler/framework/plugins/imagelocality/image_locality.go:
```diff
@@ -41,17 +41,67 @@ type ImageLocality struct {
 	return Name
 }
 
+func (pl *ImageLocality) PreScore(ctx context.Context, cycleState *framework.CycleState, pod *v1.Pod, nodes []*v1.Node) *framework.Status {
```
done

xiaomudk replied:
done

xiaomudk commented:
@kerthcet Sorry to keep you waiting.
Thank you very much for your review!

xiaomudk commented:
@kerthcet I have improved it again, please review, thank you.

kerthcet commented on pkg/scheduler/framework/plugins/imagelocality/image_locality.go:
```diff
@@ -94,10 +158,10 @@ func calculatePriority(sumScores int64, numContainers int) int64 {
 // Each image receives a raw score of its size, scaled by scaledImageScore. The raw scores are later used to calculate
 // the final score. Note that the init containers are not considered for it's rare for users to deploy huge init containers.
-func sumImageScores(nodeInfo *framework.NodeInfo, containers []v1.Container, totalNumNodes int) int64 {
+func sumImageScores(nodeInfo *framework.NodeInfo, containers sets.Set[string], totalNumNodes int) int64 {
```
Rename to `imageNames`, better?

xiaomudk replied:
done

kerthcet commented on pkg/scheduler/framework/plugins/imagelocality/image_locality_test.go:
```diff
@@ -327,6 +367,55 @@ func TestImageLocalityPriority(t *testing.T) {
+			expectedList: []framework.NodeScore{{Name: "node1", Score: 0}, {Name: "node2", Score: 0}},
+			name:         "pod with multiple small images and all ImagePullPolicy set to Always",
+		},
+		{
```
Is this testcase's result the same before or after the PR, if so, I think we should add a better one. Like before we prefer node1, after the PR, then node2. I didn't test it myself, seems the same result.

xiaomudk replied:
Deleted the duplicate testcase.

kerthcet reviewed (commented):
Totally LGTM, some suggestions left @xiaomudk

alculquicondor commented:
please ping me when it's lgtmed and squashed

kerthcet commented on pkg/scheduler/framework/plugins/imagelocality/image_locality_test.go:
```diff
@@ -327,6 +367,41 @@ func TestImageLocalityPriority(t *testing.T) {
+
+			// Node1
+			// Image: gcr.io/600:latest 600MB, gcr.io/900:latest 900MB
+			// Score(no PullAlways): 100 * (300M * 2/3 + 2000M * 1/3 + 0  - 23M) / (1000M * 3 - 23M) = 28
+			// Score: 100 * (300M * 2/3 + 0  - 23M) / (1000M * 3 - 23M) = 5
```
```suggestion
			// Score(not support PullAlways): 100 * (300M * 2/3 + 2000M * 1/3 + 0  - 23M) / (1000M * 3 - 23M) = 28
			// Score: 100 * (300M * 2/3 + 0 + 0 - 23M) / (1000M * 3 - 23M) = 5
```

kerthcet reviewed (commented):
One nit, please squash with the new commit. Thanks.

xiaomudk commented:
@alculquicondor @kerthcet Squashed and ping.

kerthcet commented:
Thanks @xiaomudk for your excellent work. /lgtm /hold
/assign @alculquicondor for another round of review.

By the way, we may have a follow up on this about considering initContainer in scoring, this can be another PR, would you like to follow this :)

xiaomudk commented:
@kerthcet We should open an issue first to discuss the details, WDYT?

kerthcet commented:
Yeah, plz go ahead.
/lgtm
/hold
trigger again.

sftim commented:
We might prefer to score Pods that always pull images differently from those that don't. However, a skip is placing too little weight on the likelihood that the node has the right layer cached.

alculquicondor commented:
/remove-kind bug
/approve cancel
/hold for discussion in https://github.com/kubernetes/kubernetes/issues/110149

sftim commented:
Based on https://github.com/kubernetes/kubernetes/pull/110382#issuecomment-1424856143
/hold
